File: kindavm/consumer.py
"""Consumer Control HID functionality (media keys, volume, brightness)."""

from .hid import send_report

# Consumer control button bits (Report ID 2, Byte 1)
VOLUME_UP = 0x01
VOLUME_DOWN = 0x02
MUTE = 0x04
PLAY_PAUSE = 0x08
NEXT_TRACK = 0x10
PREV_TRACK = 0x20
STOP = 0x40
MAIL = 0x80

# Consumer control button bits (Report ID 2, Byte 2)
BRIGHTNESS_UP = 0x01
BRIGHTNESS_DOWN = 0x02
AC_SEARCH = 0x04
AC_HOME = 0x08
AC_BACK = 0x10
AC_FORWARD = 0x20
AC_STOP = 0x40
AC_REFRESH = 0x80

# Consumer control button bits (Report ID 2, Byte 3)
EJECT = 0x01


def send_consumer_key(byte1: int = 0, byte2: int = 0, byte3: int = 0) -> None:
    """Send a consumer control key press and release.

    Args:
        byte1: First byte of consumer control bits (volume, media controls)
        byte2: Second byte of consumer control bits (brightness, browser)
        byte3: Third byte of consumer control bits (eject)
    """
    # Press (Report ID 2 + 3 bytes = 4 bytes total)
    report = bytes([0x02, byte1, byte2, byte3])
    send_report(report, delay_ms=10)

    # Release (Report ID 2 + 3 bytes = 4 bytes total)
    report = bytes([0x02, 0x00, 0x00, 0x00])
    send_report(report, delay_ms=10)


def volume_up() -> None:
    """Increase volume."""
    send_consumer_key(byte1=VOLUME_UP)


def volume_down() -> None:
    """Decrease volume."""
    send_consumer_key(byte1=VOLUME_DOWN)


def mute() -> None:
    """Toggle mute."""
    send_consumer_key(byte1=MUTE)


def play_pause() -> None:
    """Toggle play/pause."""
    send_consumer_key(byte1=PLAY_PAUSE)


def next_track() -> None:
    """Skip to next track."""
    send_consumer_key(byte1=NEXT_TRACK)


def prev_track() -> None:
    """Skip to previous track."""
    send_consumer_key(byte1=PREV_TRACK)


def stop() -> None:
    """Stop playback."""
    send_consumer_key(byte1=STOP)


def brightness_up() -> None:
    """Increase brightness."""
    send_consumer_key(byte2=BRIGHTNESS_UP)


def brightness_down() -> None:
    """Decrease brightness."""
    send_consumer_key(byte2=BRIGHTNESS_DOWN)


def eject() -> None:
    """Eject media."""
    send_consumer_key(byte3=EJECT)

File: internal/hid/consumer.go
package hid

// Consumer control bit positions for Report ID 2
// Byte 1: Media controls
const (
	ConsumerVolumeUp      = 0x01 // bit 0
	ConsumerVolumeDown    = 0x02 // bit 1
	ConsumerMute          = 0x04 // bit 2
	ConsumerPlayPause     = 0x08 // bit 3
	ConsumerNextTrack     = 0x10 // bit 4
	ConsumerPreviousTrack = 0x20 // bit 5
	ConsumerStop          = 0x40 // bit 6
	ConsumerEmail         = 0x80 // bit 7
)

// Byte 2: Brightness and browser controls
const (
	ConsumerBrightnessUp   = 0x01 // bit 0
	ConsumerBrightnessDown = 0x02 // bit 1
	ConsumerACSearch       = 0x04 // bit 2
	ConsumerACHome         = 0x08 // bit 3
	ConsumerACBack         = 0x10 // bit 4
	ConsumerACForward      = 0x20 // bit 5
	ConsumerACStop         = 0x40 // bit 6
	ConsumerACRefresh      = 0x80 // bit 7
)

// Byte 3: Additional controls
const (
	ConsumerEject = 0x01 // bit 0
)

// Consumer represents a HID consumer control interface
type Consumer struct {
	device *Device
}

// NewConsumer creates a new consumer control interface
func NewConsumer(device *Device) *Consumer {
	return &Consumer{device: device}
}

// SendConsumerReport sends a consumer control report
// Report format (4 bytes):
//
//	Byte 0: Report ID (0x02)
//	Byte 1: Media control bits
//	Byte 2: Brightness/browser control bits
//	Byte 3: Additional control bits
func (c *Consumer) SendConsumerReport(byte1, byte2, byte3 byte) error {
	report := []byte{
		0x02,  // Report ID for consumer control
		byte1, // Media controls
		byte2, // Brightness/browser controls
		byte3, // Additional controls
	}
	return c.device.SendReport(report, DefaultDelayMS)
}

// SendConsumerKey sends a consumer key press and release
func (c *Consumer) SendConsumerKey(byte1, byte2, byte3 byte) error {
	// Press
	if err := c.SendConsumerReport(byte1, byte2, byte3); err != nil {
		return err
	}
	// Release
	return c.SendConsumerReport(0x00, 0x00, 0x00)
}

// VolumeUp sends a volume up key press
func (c *Consumer) VolumeUp() error {
	return c.SendConsumerKey(ConsumerVolumeUp, 0x00, 0x00)
}

// VolumeDown sends a volume down key press
func (c *Consumer) VolumeDown() error {
	return c.SendConsumerKey(ConsumerVolumeDown, 0x00, 0x00)
}

// BrightnessUp sends a brightness up key press
func (c *Consumer) BrightnessUp() error {
	return c.SendConsumerKey(0x00, ConsumerBrightnessUp, 0x00)
}

// BrightnessDown sends a brightness down key press
func (c *Consumer) BrightnessDown() error {
	return c.SendConsumerKey(0x00, ConsumerBrightnessDown, 0x00)
}
